### DISCUSSION TITLE
Intercept on Thread

### DISCUSSION THREAD

**Discussion by waihangcheung** [btraceio/btrace#542] [General]
Hi, 

I've been trying to intercept on Thread on application startup, and got no luck. I know that there are similar samples provided in the project  but is there any specific config i should set? Or am i missing anything?

`-javaagent:${BTRACE_HOME}/btrace-agent.jar=script=Test.class,noServer=true,scriptOutputFile=stdout.log`

My btrace script
```
@BTrace
public class Test
{
    @OnMethod(clazz = "java.lang.Thread", method = "start")
    public static void onNewThread()
    {
        println("New thread spawned");
    }
}
```

My sample application 
```
public class Main
{
    public static void main(String[] args) throws InterruptedException
    {
        Thread thread = new Thread()
        {
            public void run()
            {
                System.out.println("Thread Running");
            }
        };

        thread.start();
        thread.join();
        System.out.println("Thread done");
    }
}
```

**Comment by jbachorik** [comment]
BTrace needs to exclude any classes it requires to run the core functionality from the instrumentation candidates to prevent possible circularity errors during the invokedynamic bootstrap :(

Currently, the filter is a bit overzealous because I didn't have the time to fine-tune the excludes and as such all `java.lang.*` classes will not be instrumented.

Can you intercept a `Runnable` instead?

**Reply by waihangcheung** [reply]
I guess it's the same restriction for `java.util.concurrent.*`?

**Comment by waihangcheung** [comment]
No luck, Runnable is under `java.lang.Runnable`

**Reply by jbachorik** [reply]
It's also an interface :)
You would need something like `clazz = "+java.lang.Runnable"` to instrument all implementations.
That might be actually quite a lot of classes.

**Comment by waihangcheung** [comment]
`clazz = "+java.lang.Runnable"` does work. 

I want to narrow it down to Kind.CALL, doesn't seem to work

```
    @OnMethod(clazz = "TargetClass", method = "targetMethod",
              location = @Location(value = Kind.CALL, clazz = "+java.lang.Runnable", method = "run"))
```

**Reply by jbachorik** [reply]
You could try patching https://github.com/btraceio/btrace/blob/46cf07f95738429e7fa2d130321310a10b00b179/btrace-instr/src/main/java/org/openjdk/btrace/instr/ClassFilter.java#L189 to exclude `java.lang.Thread` and test whether the agent starts normally and the method is properly traced. 

I am away from my computer for next 5 days so there is no chance I might try this on my own.

**Reply by waihangcheung** [reply]
commented out `java/lang/` & `java/util/` , it seems to work fine. Thanks

Need to also enable `startupRetransform=true`

**Reply by jbachorik** [reply]
Modifications to the sensitive classes list will take more effort. 
The problem is with invoke dynamic that is necessary for BTrace runnning on JDK 15 and newer - the  linking of a dynamic call involves quite a lot of JDK classes and if any of them happen to be instrumented as well the whole JVM deadlocks.
